File: jdroid-android-core/src/main/java/com/jdroid/android/analytics/CoreAnalyticsTracker.java
package com.jdroid.android.analytics;

import android.app.Activity;
import android.net.Uri;
import android.os.Bundle;

import com.google.android.play.core.splitinstall.SplitInstallSessionState;
import com.jdroid.android.social.SocialAction;
import com.jdroid.java.analytics.AnalyticsTracker;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

public interface CoreAnalyticsTracker extends AnalyticsTracker {

	// Error handling

	void trackHandledException(Throwable throwable, List<String> tags);

	void trackErrorLog(@NonNull String message);

	void trackErrorCustomKey(@NonNull String key, @NonNull Object value);

	// Activity/fragment life cycle

	void onFirstActivityCreate(Activity activity);

	void onActivityCreate(Activity activity, @Nullable Bundle savedInstanceState);

	void onActivityStart(Activity activity, @Nullable String referrer, @Nullable Object data);

	void onActivityResume(Activity activity);

	void onActivityPause(Activity activity);

	void onActivityStop(Activity activity);

	void onActivityDestroy(Activity activity);

	void onFragmentStart(String screenViewName);

	// Notifications

	void trackNotificationDisplayed(String notificationName);

	void trackNotificationOpened(String notificationName);

	// Feedback

	void trackEnjoyingApp(Boolean enjoying);

	void trackRateOnGooglePlay(Boolean rate);

	void trackGiveFeedback(Boolean feedback);

	// Widgets

	void trackWidgetAdded(String widgetName);

	void trackWidgetRemoved(String widgetName);
	
	// Split Install

	void trackSplitInstallStatus(String moduleName, SplitInstallSessionState state);

	void trackSplitInstallUninstalled(String moduleName);

	// More

	void trackUriOpened(String screenName, Uri uri, String referrer);

	void trackSocialInteraction(String network, SocialAction socialAction, String socialTarget);

}

File: jdroid-android-core/src/main/java/com/jdroid/android/analytics/CoreAnalyticsTracker.kt
package com.jdroid.android.analytics

import android.app.Activity
import android.net.Uri
import android.os.Bundle

import com.google.android.play.core.splitinstall.SplitInstallSessionState
import com.jdroid.android.social.SocialAction
import com.jdroid.java.analytics.AnalyticsTracker

interface CoreAnalyticsTracker : AnalyticsTracker {

    // Error handling

    fun trackHandledException(throwable: Throwable, tags: List<String>)

    fun trackErrorLog(message: String)

    fun trackErrorCustomKey(key: String, value: Any)

    // Activity/fragment life cycle

    fun onFirstActivityCreate(activity: Activity)

    fun onActivityCreate(activity: Activity, savedInstanceState: Bundle?)

    fun onActivityStart(activity: Activity, referrer: String?, data: Any?)

    fun onActivityResume(activity: Activity)

    fun onActivityPause(activity: Activity)

    fun onActivityStop(activity: Activity)

    fun onActivityDestroy(activity: Activity)

    fun onFragmentStart(screenViewName: String)

    // Notifications

    fun trackNotificationDisplayed(notificationName: String)

    fun trackNotificationOpened(notificationName: String)

    // Feedback

    fun trackEnjoyingApp(enjoying: Boolean)

    fun trackRateOnGooglePlay(rate: Boolean)

    fun trackGiveFeedback(feedback: Boolean)

    // Widgets

    fun trackWidgetAdded(widgetName: String)

    fun trackWidgetRemoved(widgetName: String)

    // Split Install

    fun trackSplitInstallStatus(moduleName: String, state: SplitInstallSessionState)

    fun trackSplitInstallDeferredUninstall(moduleName: String)

    // More

    fun trackUriOpened(screenName: String, uri: Uri, referrer: String)

    fun trackSocialInteraction(network: String?, socialAction: SocialAction, socialTarget: String)
}
